rfcs/20220404-troubleshooting-helpers/README.md

```diff
@@ -0,0 +1,304 @@
+be obtained from status, as:
+- the primitive descriptor was created successfully, even though the
+  dispatched implementation is not the one expected by the user.
+- each implementation that was not dispatched might have a different
+  reason for not dispatching
+
+For the reasons above, we propose to leave the statuses as they
+currently are, so that they align with std::exception, and can be
+easily mapped from C to C++ API.
+
+
+## 2.2 Release mode verbose vs debug mode verbose
+Some debugging information might impact the performance of the
+library, in particular primitive creation and execution time.
+Currently, we already have such debugging facilities all over the
+place, and they are mostly controlled with dedicated macros, that are
+implementation specific (to name a few, we have `GEN_CONV_DEBUG`,
+`TR_DEBUG`, `DEBUGPRINT`).
+
+Here we propose to unify those macros. The main benefit is, for
+developers, that the mechanism to print and debug would be the same
+across implementations.
+
+This macro should be independent of `NDEBUG` as we might need to use
+this logging information on the release build. We propose to call this
+unified macro `ONEDNN_DEVEL_MODE`, and it should have a matching cmake
+option.
+
+We recommend that all messages related to internal logic go under
+`ONEDNN_DEVEL_MODE`. Only messages related to user visible APIs which
+are actionable by user should live outside `ONEDNN_DEVEL_MODE`.
+
+## 2.3 Verbose levels vs verbose flags
+
+Currently, we have some documented information messages we provide
+(primitive creation/execution time), and some undocumented
+(e.g. runtimes info and implementation details).  For the undocumented
+ones, we don't have alignment on which level they should print (some
+print at level 2 and above, other print at level 5 and above).
+
+The current verbose levels were designed with the assumption that each
+level of information was inclusive of the previous ones. If we want to
+increase the number of categories of information to print
+(e.g. dispatch info, perf, error info, debug info, ...), it will
+quickly bloat the verbose output, which can become an issue when using
+verbose for long running workloads.  More importantly, verbose
+messages might impact runtime (e.g. getting perf on GPU), so getting
+some information sometimes impacts runtime.
+
+We propose here to use flags in order to limit verbose bloat, and
+allow printing messages without impacting runtime when possible. This
+could take two forms:
+- a list of message types, like `ERROR,PROFILE,DISPATCH,DEVINFO`...
+- a bit mask encoded in an integer, -1 being all messages are
+  displayed. So we would have for example ERROR=1, INFO=2, DISPATCH=4,
+  ... and the user would have to pass the sum of all diagnostics they
+  want (e.g. for `ERROR` and `DISPATCH`, they would have to pass
+  `ONEDNN_VERBOSE=5`).
+
+We would recommend going with the first option as it is simpler on the
```

Comment:
These message types look similar to typical logging levels. For convenience they are usually mapped to numerical values, e.g. in [minloglevel in glog](https://github.com/google/glog#setting-flags) or in [python](https://docs.python.org/3/library/logging.html#logging-levels). We could try to stick to the same design for simplicity (typing `DNNL_VERBOSE=5` is a bit faster comparing with `DNNL_VERBOSE=devinfo` especially if it's intended for developers).

Reply:
Here the primary targeted users are the oneDNN users, not the oneDNN devs.
This is why I would favor simple names, to numbers that have to be looked up in documentation for casual users (In particular, I would expect users to build oneDNN with dev mode off, and use `ONEDNN_VERBOSE=ALL` when debugging their integration code, or `ONEDNN_VERBOSE=PROFILE` when looking at performance).

For oneDNN devs, we can have number aliases to simplify if there is value (it will likely be something like a bitfield though, similar to `ZE_DEBUG`). Does it sound reasonable to you?

Reply:
Yeah, this is what I had in mind: supporting both, names and numbers. The only issue is that you mentioned we want to reserve `1` and `2`, I'm not sure if we can fit it into the bitmasks approach.